File: src/Domain/Entities/IpNote.cs
namespace Ntk.Note.IP.Domain.Entities;

/// <summary>
/// User note attached to an IP address (IPNote.ir core aggregate).
/// </summary>
public class IpNote : BaseAuditableEntity
{
    public string Address { get; set; } = string.Empty;

    public string? Title { get; set; }

    public string? Body { get; set; }

    public string? Tags { get; set; }

    /// <summary>Client-reported capture time (timezone-aware).</summary>
    public DateTimeOffset? NotedAtClient { get; set; }

    public string? ClientTimezone { get; set; }

    public string? LocalIpAddress { get; set; }

    public string? CountryCode { get; set; }

    public string? Region { get; set; }

    public string? City { get; set; }

    public string? Isp { get; set; }

    public string? Asn { get; set; }

    /// <summary>Short device label for list views (e.g. Chrome · Windows).</summary>
    public string? DeviceLabel { get; set; }

    /// <summary>Full device info JSON captured at note time.</summary>
    public string? DeviceInfoJson { get; set; }

    /// <summary>Full IP lookup snapshot JSON (geo, ASN, ISP, etc.).</summary>
    public string? IpSnapshotJson { get; set; }

    public bool IsSoftDeleted { get; set; }
}

File: src/Mobile/ntk_note_ip_app/lib/domain/entities/ip_note.dart
class IpNote {
  const IpNote({
    required this.id,
    required this.address,
    required this.created,
    this.title,
    this.body,
    this.tags,
    this.notedAtClient,
    this.clientTimezone,
    this.localIpAddress,
    this.countryCode,
    this.region,
    this.city,
    this.isp,
    this.asn,
    this.deviceLabel,
  });

  final int id;
  final String address;
  final String created;
  final String? title;
  final String? body;
  final String? tags;
  final String? notedAtClient;
  final String? clientTimezone;
  final String? localIpAddress;
  final String? countryCode;
  final String? region;
  final String? city;
  final String? isp;
  final String? asn;
  final String? deviceLabel;

  factory IpNote.fromJson(Map<String, dynamic> json) {
    return IpNote(
      id: _toInt(json['id']),
      address: json['address']?.toString() ?? '',
      created: json['created']?.toString() ?? '',
      title: json['title']?.toString(),
      body: json['body']?.toString(),
      tags: json['tags']?.toString(),
      notedAtClient: json['notedAtClient']?.toString(),
      clientTimezone: json['clientTimezone']?.toString(),
      localIpAddress: json['localIpAddress']?.toString(),
      countryCode: json['countryCode']?.toString(),
      region: json['region']?.toString(),
      city: json['city']?.toString(),
      isp: json['isp']?.toString(),
      asn: json['asn']?.toString(),
      deviceLabel: json['deviceLabel']?.toString(),
    );
  }

  List<String> get tagList {
    if (tags == null || tags!.trim().isEmpty) {
      return const [];
    }

    return tags!
        .split(',')
        .map((t) => t.trim())
        .where((t) => t.isNotEmpty)
        .toList();
  }

  String get locationLine {
    return [
      city,
      region,
      countryCode?.toUpperCase(),
    ].whereType<String>().where((v) => v.isNotEmpty).join(' · ');
  }

  String get notedWhen => notedAtClient ?? created;
}

int _toInt(Object? value) {
  if (value is int) {
    return value;
  }

  return int.tryParse(value?.toString() ?? '') ?? 0;
}
